가이드라인_피해 타입 검사와 진행도 수정,조회 기능 구현

jinseohyun1228 opened the pull request:
## 피해 타입 검사와 진행도 수정,조회 기능 구현 

##### 반영 브랜치 : `feat/34-survey-Embedded` -> `week7`
#### 관련 이슈 번호 : #35  #34 

### 🛠️ 작업 사항
- [x] 기능 추가

## ️️ 📝 설명
가이드라인기능을 위한 타입검사와, 진행도 변경, 조회 기능을 추가했습니다.

## 👨‍💻 리뷰 요구사항
### 진행도 엔티티 구성 방법
1. 진행도 타입별 엔티티를 따로 둔다.
진행도 엔티티 인터페이스를 두고, 구현체로 A, B, C, D 엔티티를 만듬

다형성을 활용해 타입별로 메서드를 호출했을때 내부로직을 신경쓰지 않아도 될 것 같습니다.

2. 진행도 범용 엔티티를 둔다.
현재, 타입별로 엔티티를 다르게 두는게 어려워 전략 객체를 사용하게 되었습니다. 

### 앤드포인트
아래 엔드포인트가 적절할까요..? 

## API 요청
### 타입
#### 타입 검사
기존에 검사한 경우에 에러 발생

- POST  `/api/v1/progress/type`
```
{
    "is_insured": true,                        //보험여부
    "is_contract_terminated" : true   //계약해지여부
}
```
```
{
    "type_name": "D타입",
    "description": "보험 가입 / 계약 해지"
}
```
#### 타입 조회
- GET  `/api/v1/progress/type`
```
{
    "type_name": "D타입",
    "description": "보험 가입 / 계약 해지"
}
```
### 진행도 
#### 진행도 업데이트(완료/미완료)
- 완료일 경우 -> PODT `api/v1/progress/{main}/{sub}`

- 미완료일 경우 -> DELETE `api/v1/progress/{main}/{sub}`


#### 진행도 조회 
- 진행도가 메인/서브 단계로 되어있어서, 전체 진행도, 특정 메인 단계 진행도, 특정 서브 단계 진행도로 구현했습니다.

- 전체 진행도 : GET / `/api/v1/progress`  
```
{
    "progress_type": {
        "type_name": "A타입",
        "description": "보험 미가입 / 계약 미해지"
    },
    "main_step_list": [
        {
            "main_step_name": "진행도1번",
            "completion": false,
            "sub_step_list": [
                {
                    "name": "sub_step1",
                    "completion": false
                }           
            ]
        }
    ]
}
```

- 특정 메인 단계 진행도 : GET / `/api/v1/progress/{main}` 
```
{
    "main_step_name": "진행도1번",
    "completion": false,
    "sub_step_list": [
        {
            "name": "sub_step1",
            "completion": false
        },
        {
            "name": "sub_step2",
            "completion": false
        }
    ]
}
```
- 특정 서브 단계 진행도 : GET / `/api/v1/progress/{main}/{sub}`
```
{
    "name": "sub_step1",
    "completion": false
}
```

jinseohyun1228 committed:
feat(Member): 멤버 엔티티에 ProgressType 필드 추가, ProgressType 등록 메서드와 ProgressType가 있는지 조회 메서드

jinseohyun1228 committed:
feat(ProgressType): 진행타입 ENUM 클래스 구현, Json 객체 반환을 위한 ProgressTypeJson 내부 클래스 구현

jinseohyun1228 committed:
feat(ProgressTypeRequestDto): 가이드라인 전, 타입 검사를 위한 요청 객체 구현

jinseohyun1228 committed:
feat(ProgressService): 진행도 검사 구현

jinseohyun1228 committed:
feat(ErrorCode): 진행도 검사 중, 요청 객체가 잘못되었거나, 이미 타입 검사를 한경우의 에러 코드 추가

jinseohyun1228 committed:
feat: 진행도 타입별 엔티티와 레파지토리 구현

jinseohyun1228 committed:
Revert "feat(ErrorCode): 진행도 검사 중, 요청 객체가 잘못되었거나, 이미 타입 검사를 한경우의 에러 코드 추가"

This reverts commit f82037ec4a8705841ec269e17e4b532c73d544f0.

jinseohyun1228 committed:
Revert "feat(ProgressService): 진행도 검사 구현"

This reverts commit f6342eec20fed2dfe9919de36bbf772707aade81.

jinseohyun1228 committed:
revert : 타입검사까지 작업 사항을 되돌림

jinseohyun1228 committed:
feat : 가이드라인 엔티티 구현

jinseohyun1228 committed:
feat : 유저의 타입 조회

jinseohyun1228 committed:
feat : 타입 검사를 하지 않고 조회할 경우의 에러 코드 추가

jinseohyun1228 committed:
feat : 세부 타입 검사 로직 추가

jinseohyun1228 committed:
feat : 메인 단계 ENum 구현

jinseohyun1228 committed:
feat : 진행도 엔티티와 리파지토리 구현

jinseohyun1228 committed:
feat : 멤버에 진행도 매핑

jinseohyun1228 committed:
feat : 진행도 중 필요한 에러 코드 추가

jinseohyun1228 committed:
feat : 진행도 업데이트와 타입 검사시 진행도 엔티티가 생성될 수 있도록 수정

jinseohyun1228 committed:
feat(ProgressController.java) : 진행도 조회 API 구현
1. 진행도 전체 조회
2. 메인 단계의 진행도 조회
3. 세부 단계의 진행 조회

jinseohyun1228 committed:
feat(ProgressStrategy) : 메인단계나, 서브단계 정보에 따라 필요한 객체의 값을 가져올 수 있도록 변경

jinseohyun1228 committed:
feat(ProgressStrategy) : 엔티티 내부에서 전략객체를 호출하도록 변경

jinseohyun1228 committed:
feat : 메인단계와 서브 단계에 대한 Enum 클래스와 컨버터 생성

jinseohyun1228 committed:
feat : 필요한 에러코드 구현

jinseohyun1228 committed:
feat : 진행도 내부의 메인 단계 구현

jinseohyun1228 committed:
feat : 진행도 조회시 필요한 응답 객체 생성

jinseohyun1228 committed:
feat : 멤버가 프로그래스와 양방향 관계 설정

jinseohyun1228 committed:
remove : 필요없는 파일 삭제

jinseohyun1228 committed:
refactor: 메서드명을 조금 더 구체적으로 변경

jinseohyun1228 committed:
refactor: 삭제와 업데이트 메서드 명 변경

jinseohyun1228 committed:
refactor: 삭제와 업데이트 메서드 명 변경

jinseohyun1228 committed:
feat : 업데이트 문 추가

jinseohyun1228 committed:
merge : week7과 병합

J-1ac reviewed (approved):
고봉밥 같은 구현양 ㄷㄷㄷ 솔선수범 팀장 칭찬해~

jinseohyun1228 commented:
앗~! 이거 프론트 요청으로 응답 객체 수정해야해욥

jinseohyun1228 committed:
merge: 최신 반영 사항 병합

jinseohyun1228 committed:
fix : 에러 설명이 다르게 매칭된 부분 수정

jinseohyun1228 committed:
fix(SubStepDto) : SubStepDto 추가

jinseohyun1228 committed:
fix : 진행도 전략 객체를 서비스에서 호출하도록 변경

jinseohyun1228 committed:
feat(SubStepResponseDto) : 생성시 SubStepDto를 전달받아서 생성

jinseohyun1228 committed:
feat(ProgressDetailsResponseDto): 생성시 메인 리스트를 받아서 생성

jinseohyun1228 committed:
feat : MainStep의 메서드명과 반환값 변경
메서드명: getSubStepAllForResponse -> getSubStepAll
반환값 : SubStepResponseDto -> SubStepDto

jinseohyun1228 committed:
merge : week7과 충돌 해결 병합
